src/proj.rs

```diff
@@ -0,0 +1,791 @@
+               interpolation: Interpolation,
+               default: P,
+               out: &mut Image<P>
+              )
+    where
+    P: Pixel + Send + Sync + 'static,
+    <P as Pixel>::Subpixel: Send + Sync,
+    <P as Pixel>::Subpixel: ValueInto<f32> + Clamp<f32> + Sync,
+{
+    let nn = |x,y| nearest(image, x, y, default);
+    let bl = |x,y| interpolate(image, x, y, default);
+    let wt = |x,y| homography.warp_t(x,y);
+    let wa = |x,y| homography.warp_a(x,y);
+    let wp = |x,y| homography.warp_p(x,y);
+    use TransformationClass as TC;
+    use Interpolation as I;
+
+    match (interpolation, homography.class) {
+        (I::Nearest,  TC::Translation) => warp_inner(out, wt, nn),
+        (I::Nearest,  TC::Affine)      => warp_inner(out, wa, nn),
+        (I::Nearest,  TC::Projection)  => warp_inner(out, wp, nn),
+        (I::Bilinear, TC::Translation) => warp_inner(out, wt, bl),
+        (I::Bilinear, TC::Affine)      => warp_inner(out, wa, bl),
+        (I::Bilinear, TC::Projection)  => warp_inner(out, wp, bl),
+    }
+}
+
+/// Transforms input `image` into output image of the same size.
+/// Fm `mapping` is the coordinate mapping function, mapping from out to in.
+pub fn warp_new_with<P, Fm>(mapping: Fm,
+                   image: &Image<P>,
+                   interpolation: Interpolation,
+                   default: P
+                  ) -> Image<P> 
+    where
+    Fm: Fn(f32, f32) -> (f32, f32) + Sync + Send,
+    P: Pixel + Send + Sync + 'static,
+    <P as Pixel>::Subpixel: Send + Sync,
+    <P as Pixel>::Subpixel: ValueInto<f32> + Clamp<f32>,
+{
+    let (width, height) = image.dimensions();
+    let mut out = ImageBuffer::new(width, height);
+    warp_with(mapping, image, interpolation, default, &mut out);
+
+    out
+}
+
+/// Warp image with custom function.
+/// This enables the user to define a custom mapping such as a wave pattern:
+/// ```
+/// use image::{ImageBuffer, Luma};
+/// use imageproc::utils::gray_bench_image;
+/// use imageproc::proj::*;
+///
+/// let img = gray_bench_image(300, 300);
+/// let mut out = ImageBuffer::new(300, 300);
+/// warp_with(|x,y| (x, y+(x/30.0).sin()), &img, Interpolation::Nearest, Luma([0u8]), &mut
+/// out);
+/// ```
+pub fn warp_with<P, Fm>(mapping: Fm,
+                   image: &Image<P>,
+                   interpolation: Interpolation,
+                   default: P,
+                   out: &mut Image<P>
+                  )
+    where
+    Fm: Fn(f32, f32) -> (f32, f32) + Send + Sync,
+    P: Pixel + Send + Sync + 'static,
+    <P as Pixel>::Subpixel: Send + Sync,
+    <P as Pixel>::Subpixel: ValueInto<f32> + Clamp<f32>,
+
+{
+    let nn = |x,y| nearest(image, x, y, default);
+    let bl = |x,y| interpolate(image, x, y, default);
+    use Interpolation as I;
+
+    match interpolation {
+        I::Nearest => warp_inner(out, mapping, nn),
+        I::Bilinear => warp_inner(out, mapping, bl),
+    }
+}
+
+fn warp_inner<P,Fc,Fi>(
+    out: &mut Image<P>,
+    coord_tf: Fc,
```

Comment:
This parameter is called `mapping` in the previous function, which is more understandable.